File: arkiv/1.Semester/TFY4145 - Mekanisk Fysikk/Øvinger + LF (Høst 2015)/Øving 10/animasjon.py
#! /usr/bin/env python3
# -*- coding: UTF-8 -*- 

from numpy import linspace,zeros,sin,sqrt,pi
from matplotlib.pyplot import figure,axes,plot,show,xlabel,ylabel,title
from matplotlib.animation import FuncAnimation

#   Funksjon som beregner strengens utsving (innkommende + reflektert bolge for x<0,
#   og transmittert bolge for x>0)
def set_utsving(x,y,A,B,C,k1,k2,w,t,N):
    # For x<0
    y[0:N]=A*sin(k1*x[0:N]-w*t)+B*sin(k1*x[0:N]+w*t)
    # For x>0
    y[N:2*N]=C*sin(k2*x[N:2*N]-w*t)


def svingende_streng(mu_1,mu_2,f,S,A):
    v_1=sqrt(S/mu_1)            # Bolgehastighet x<0
    v_2=sqrt(S/mu_2)            # Bolgehastighet x>0
    w=2*pi*f                    # Vinkelfrekvens
    k_1=w/v_1                   # Bolgetall x<0
    k_2=w/v_2                   # Bolgetall x>0
    B=A*(sqrt(mu_2)-sqrt(mu_1))/(sqrt(mu_1)+sqrt(mu_2)) # Amplitude reflektert bolge
    C=A*2.0*sqrt(mu_1)/(sqrt(mu_1)+sqrt(mu_2))          # Amplitude transmittert bolge
    N=400                       # Halve antall punkter i x-retning
    M=400                       # 
    k_min=k_1
    if(k_2<k_1):
        k_min=k_2
    x_max=2*2*pi/k_min          # Storste x-verdi
    y_max=1.25*(abs(A)+abs(B))  #
    x=linspace(-x_max,x_max,num=2*N)    # Tabell med 2N x-verdier mellom -x_max og x_max
    y=zeros(2*N)                        # Tabell med 2N "nuller"

    # Animere strengen
    figur=figure('Figur')               
    akse=axes(xlim=(-pi,pi),ylim=(-y_max,y_max))
    title('Transmisjon og refleksjon i x=0 av harmonisk bolge paa streng')
    xlabel('posisjon '+r'$(m)$')
    ylabel('utsving  '+r'$(mm)$')
    linje, =plot([], [], lw=2)
    def initialisering():
        linje.set_data([], [])
        return linje,
    def animere(i):
        set_utsving(x,y,A,B,C,k_1,k_2,w,i*10*2*pi/(w*M),N)
        linje.set_data(x,y)
        return linje,
    animasjon=FuncAnimation(figur,animere,init_func=initialisering,frames=200,interval=50,blit=True)
    show()

if __name__=='__main__':
    # Definere parametre
    mu_1=.01    # Massetetthet for x<0
    mu_2=.09    # Massetetthet for x>0
    f=5.0       # Frekvens for innkommende bolge
    S=4.0       # Snordrag
    A=5.0       # Amplitude for innkommende bolge
    # Her kaller vi funsjonen
    svingende_streng(mu_1,mu_2,f,S,A)

File: arkiv/1.Semester/TFY4145 - Mekanisk Fysikk/Øvinger + LF (Høst 2015)/Øving 10/animasjon.m
function animasjon

clear all;

pi = 3.141592654;
% Strekk-kraft (N = kg*m/s*s):
S = 4.0;
% Massetettheter (kg/m);
mu1 = 0.010;
mu2 = 0.090;
% Bolgehastigheter (m/s):
v1 = sqrt(S/mu1);
v2 = sqrt(S/mu2);
% Frekvens (Hz):
f = 5.0;
% Vinkelfrekvens (1/s):
w = 2*pi*f;
% Dette tilsvarer en periode T = 2*pi/w = 0.2 s.
% Bolgetall (1/m):
k1 = w/v1;
k2 = w/v2;
% Amplituden til innkommende bolge (mm):
A = 5.0;
% Amplituden til reflektert bolge (mm):
B = A * (sqrt(mu2)-sqrt(mu1))/(sqrt(mu2)+sqrt(mu1));
% Amplituden til transmittert bolge (mm):
C = A * 2.0 * sqrt(mu1)/(sqrt(mu2)+sqrt(mu1));
% Det halve av antall punkter langs x-aksen:
N = 400;
% Antall tidssteg:
M = 400;
% Tabeller for x- og y-verdier, begge med 2*N elementer.
% Velger x-intervallet slik at vi faar minst 2 bolgelengder paa
% begge sider av x=0
kmin = min(k1,k2);
xmax = 2*2*pi/kmin;
ymax = 1.25*(abs(A) + abs(B));
x = linspace(-xmax,xmax,2*N);
y = linspace(-ymax,ymax,2*N);

% Animasjon av refleksjon/transmisjon i x = 0:
p = plot(x,y,'-','EraseMode','xor'); 
axis([-xmax xmax -ymax ymax])
hold on
xlabel('posisjon (m)');
ylabel('utsving (mm)');
title('Transmisjon og refleksjon i x = 0 av harmonisk bølge på streng');
for i=1:M
    % Velger tidssteg slik at animasjonen varer i 10 hele perioder:
    t = i*10*2*pi/(w*M);
    % For j = 1:N er x<0, og der er y = y_inn + y_refl:
    for j = 1:N
        y(j) = A*sin(k1*x(j)-w*t) + B*sin(k1*x(j)+w*t);
%%      kun innkommende:
%%      y(j) = A*sin(k1*x(j)-w*t);
%%      kun reflektert:
%%      y(j) = B*sin(k1*x(j)+w*t);
    end;
    % For j = N+1:2*N er x>0, og der er y = y_trans:
    for j = N+1:2*N
        y(j) = C*sin(k2*x(j)-w*t);
    end;
    set(p,'XData',x,'YData',y) 
    % MATLAB help: drawnow flushes the event queue and updates the figure window
    drawnow    
    pause(0.050);  % Forsinker framvisningen i 0.05 s (juster hvis nodvendig)
end
